### PR TITLE
Issue 6065 - Adding ZVol as log device of seprate pool causes deadlock

### PR DESCRIPTION
**Pull request by bprotopopov**
Introduced zv_state_lock to protect internal state of zvol_state_t. 
Reverted commit 1ee159f423b5eb3c4646b0ba2dd0fb359503ba90
        Fix lock order inversion with zvol_open()... 

### Description

Introduced zv_state_lock to protect internal state of zvol_state_t and to avoid taking spa_namespace_lock (e.g. in dmu_objset_own() code path) while holding zvol_stat_lock. Refactore the code accordingly.

Reverted commit 1ee159f423b5eb3c4646b0ba2dd0fb359503ba90
        Fix lock order inversion with zvol_open()... 
as it did not account for use of zvols as vdevs. The latter use cases resulted in the lock order inversion deadlocks that involved spa_namespace_lock and bdev->bd_mutex.

### Motivation and Context

This is needed to allow use of zvols as vdevs (logs, caches, regular vdevs).

### How Has This Been Tested?

Tested use of zvols as regular vdevs, logs, and caches. Added and removed devices, exported, and re-imported pools. Performed basic I/O testing. Ran full suit of zfs-test (did not enable additional relevant test cases at this time).

### Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [X] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by bprotopopov** [comment]
Addressed commit message style issues:
error: commit subject over 50 characters
error: missing "Signed-off-by"

File: module/zfs/zvol.c [hunk 1]
```diff
@@ -38,6 +38,33 @@
...
  */
 
+/*
+ * Note on locking of zvol state structures.
+ *
+ * These structures are used to maintain internal state used to emulate block
+ * devices on top of zvols. In particular, management of device minor number
+ * operations - create, remove, rename, and set_snapdev - involves access to
+ * these structures. The zvol_state_lock is primarily used to protect the
+ * zvol_state_list,  while the zv_state_lock is used to protect the contents
```
Comment (behlendorf) [inline module/zfs/zvol.c]: The names `zvol_state_lock` and `zv_state_lock` are so close how about adding `zv->` here to make it clear it's a member of the structure.  And I think we can break this in to two sentences.

```diff
- * zvol_state_list,  while the zv_state_lock is used to protect the contents
+ * zvol_state_list.  The zv->zv_state_lock is used to protect the contents
```

File: module/zfs/zvol.c [hunk 2]
```diff
@@ -38,6 +38,33 @@
...
+ * of the zvol_state_t structures, as well as to make sure that when it the
```
Comment (behlendorf) [inline module/zfs/zvol.c]: s/when it the/when the/

File: module/zfs/zvol.c [hunk 3]
```diff
@@ -367,12 +393,16 @@ zvol_set_volsize(const char *name, uint64_t volsize)
...
+	mutex_exit(&zvol_state_lock);
 
 	if (zv == NULL || zv->zv_objset == NULL) {
 		if ((error = dmu_objset_own(name, DMU_OST_ZVOL, B_FALSE,
 		    FTAG, &os)) != 0) {
-			mutex_exit(&zvol_state_lock);
-			return (SET_ERROR(error));
+			if (zv != NULL)
+				mutex_exit(&zv->zv_state_lock);
+			return (SET_ERROR(ENXIO));
```
Comment (behlendorf) [inline module/zfs/zvol.c]: Did you find an issue here returning `error` instead of `ENXIO` unconditionally?

File: module/zfs/zvol.c [hunk 4]
```diff
@@ -1918,22 +1945,36 @@ zvol_remove_minors_impl(const char *name)
...
 	}
 	mutex_exit(&zvol_state_lock);
+
+	/*
+	 * Drop zvol_state_lock before calling zvol_free()
+	 */
+	for (zv = list_head(&free_list); zv != NULL; zv = zv_next) {
+		zv_next = list_next(&free_list, zv);
+		zvol_free(zv);
+	}
```
Comment (behlendorf) [inline module/zfs/zvol.c]: How about using the `while/list_head/list_remove` idiom here.

```
while ((zv = list_head(&free_list)) != NULL) {
        list_remove(&free_list, zv);
        zvol_free(zv);
}
```

File: module/zfs/zvol.c (hunk 1 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: Sure, will change.

File: module/zfs/zvol.c (hunk 2 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: Will do.

File: module/zfs/zvol.c (hunk 3 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: No, it looks like an oversight, will fix.

File: module/zfs/zvol.c (hunk 4 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: Sure, seems reasonable to me.

**Comment by bprotopopov** [comment]
Forced-pushed the changes.

**Comment by bprotopopov** [comment]
Noticed a small typo in the commit message, re-pushed.

**Comment by behlendorf** [comment]
Thanks for updating this.

I've updated #6128 to include your original version of this patch.  That PR additionally enables most of the test cases which depend on this functionality.  Pushed to the bots for testing, everything worked as expected for me locally.

**Comment by bprotopopov** [comment]
awesome, @behlendorf, nice to hear

**Comment by behlendorf** [comment]
The test failures were unrelated and I haven't observed any issues with this change locally.  I'm merging it to master and we'll get the test cases which depend on it enabled soon to avoid regressing.  @bprotopopov thanks for sorting this out!

**Pull request closed without merging**